corehq/apps/userreports/sql/adapter.py

```diff
@@ -48,20 +48,26 @@ def save(self, doc):
         indicator_rows = self.config.get_all_values(doc)
         if indicator_rows:
             table = self.get_table()
-            with self.engine.begin() as connection:
-                # delete all existing rows for this doc to ensure we aren't left with stale data
-                delete = table.delete(table.c.doc_id == doc['_id'])
-                connection.execute(delete)
-                for indicator_row in indicator_rows:
-                    all_values = {i.column.database_column_name: i.value for i in indicator_row}
-                    insert = table.insert().values(**all_values)
-                    try:
-                        connection.execute(insert)
-                    except IntegrityError:
-                        # Someone beat us to it. Concurrent inserts can happen
-                        # when a doc is processed by the celery rebuild task
-                        # at the same time as the pillow.
-                        pass
+            try:
+                with self.engine.begin() as connection:
+                    # delete all existing rows for this doc to ensure we aren't left with stale data
+                    delete = table.delete(table.c.doc_id == doc['_id'])
+                    connection.execute(delete)
+                    bad_saves = False
+                    for indicator_row in indicator_rows:
+                        all_values = {i.column.database_column_name: i.value for i in indicator_row}
+                        insert = table.insert().values(**all_values)
+                        try:
+                            connection.execute(insert)
+                        except (IntegrityError, InternalError):
+                            bad_saves = True
+                    if bad_saves:
+                        raise BadSaveError()
```

Comment:
From requirement 4: "[people] would prefer even partial saves as a result of (3) to get in."

I don't think the prior or current implementation would ever result in a partial save. It seems that it would either be all or nothing because of the way postgres handles transactions in conjunction with failed queries (transaction is aborted, commands ignored until end of transaction block). If we want partial saves then I think we'll need to do each iteration of the for loop in its own transaction with commit or rollback on each iteration. There is another option (save points) if all statements must be in a single transaction, but it's more complicated, and I'm not sure the performance would be any better.

Also, I'm not sure of the expected semantics of this `save` method, but since it might do a partial save it seems to be a good idea to do either one or more of the following:

1. Rename the method to `try_to_save`.
2. Add a docstring explaining that some (maybe all) rows may not be saved.
3. If there are parts of the code that would benefit from knowing if some rows were not saved, return enough information for the caller to determine what happened. Some examples of return values might be a simple true/false (all rows saved/some rows not saved), or an integer (number of rows saved), or a tuple `(total_rows, number_saved)`, or a list of rows that could not be saved. If this is done, the return value should be documented in the docstring.